pkg/database/db_factory.go

```diff
@@ -0,0 +1,300 @@
+package database
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	glog "log"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/rotisserie/eris"
+	"golang.org/x/exp/slices"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/mattn/go-sqlite3"
+	log "github.com/sirupsen/logrus"
+	"gorm.io/driver/postgres"
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+	"gorm.io/gorm/logger"
+	"gorm.io/plugin/dbresolver"
+)
+
+const (
+	SQLiteCustomDriverName = "sqlite3_custom_driver"
+)
+
+// dbFactory is the interface for all datasource creation.
+type dbFactory interface {
+	makeDbInstance() (*DbInstance, error)
+}
+
+// baseDbFactory are the common attributes for all datasources.
+type baseDbFactory struct {
+	dsnURL        url.URL
+	slowThreshold time.Duration
+	poolMax       int
+	reset         bool
+}
+
+// postgresDbFactory will make Postgres DbInstance.
+type postgresDbFactory struct {
+	baseDbFactory
+}
+
+// sqliteDbFactory will make Sqlite DbInstance.
+type sqliteDbFactory struct {
+	baseDbFactory
+}
+
+// ConnectDB will establish and return a DbInstance while also caching it in the global
+// var database.DB.
+func ConnectDB(dsn string, slowThreshold time.Duration, poolMax int, reset bool, migrate bool, artifactRoot string,
+) (*DbInstance, error) {
+	db, err := MakeDBInstance(dsn, slowThreshold, poolMax, reset, migrate, artifactRoot)
+	if err != nil {
+		return nil, err
+	}
+	// set the global DB
+	DB = db
+	return DB, nil
+}
+
+// MakeDbInstance will create a DbInstance from the parameters, without affecting the global var database.DB.
+func MakeDBInstance(
+	dsn string, slowThreshold time.Duration, poolMax int, reset bool, migrate bool, artifactRoot string,
+) (*DbInstance, error) {
+	dbFactory, err := newDbInstanceFactory(dsn, slowThreshold, poolMax, reset)
+	if err != nil {
+		return nil, err
+	}
+	db, err := dbFactory.makeDbInstance()
+	if err != nil {
+		return nil, err
+	}
+
+	if reset {
+		if err := db.reset(); err != nil {
+			db.Close()
+			return nil, err
+		}
+	}
+
+	if err := db.checkAndMigrate(migrate); err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	if err := db.createDefaultExperiment(artifactRoot); err != nil {
+		db.Close()
+		return nil, err
+	}
+
+	return db, nil
+}
+
+// newDbInstanceFactory will return the correct factory type for the datasource URI.
+func newDbInstanceFactory(
```

Comment:
it is private because not used outside this package -- the public interface is MakeDBInstance

Reply:
but again, we have a lot of different functions that call functions that call functions.

Reply:
I understood the goal of this story to attach functions to structs so they were more "object oriented" -- but at some point we have to take the DSN and decide which structs to use, which is the point of this method. It is private because other packages should not need to know about it.

Reply:
But factory pattern is a bit different. It creates something based on input parameters and return a generic interface to work with what've created. Pass dsn to factory and options and get public interface which will provide methods to work with incapsulated entities.

Reply:
so this is a `factory` pattern -> https://refactoring.guru/design-patterns/factory-method/go/example

Reply:
functions which create and return common interface. interface provide public methods to work with incapsulated logic.